gnomad_constraint/utils/constraint_basics.py

```diff
@@ -73,11 +84,87 @@ def prepare_ht_for_constraint_calculations(
     }
     if annotate_coverage:
         annotation["exome_coverage"] = ht.coverage.exomes.median
+
     ht = ht.annotate(**annotation)
     ht = add_most_severe_csq_to_tc_within_ht(ht)
     return ht
 
 
+def get_proportion_observed_by_coverage(
+    exome_ht: hl.Table,
+    context_ht: hl.Table,
+    mutation_ht: hl.Table,
+    possible_file_path: str,
+    recompute_possible: bool = False,
+    impose_high_af_cutoff_upfront: bool = True,
+    dataset: str = "gnomad",
+    af_cutoff=0.001,
+    kept_context_annotations: Tuple[str] = (
+        "context",
+        "ref",
+        "alt",
+        "methylation_level",
+        "exome_coverage",
+    ),
+    kept_exome_annotations: Tuple[str] = (
+        "context",
+        "ref",
+        "alt",
+        "methylation_level",
+        "exome_coverage",
+        "freq",
+        "pass_filters",
+    ),
+    pops: Tuple[str] = POPS,
+    grouping: Tuple[str] = ("exome_coverage"),
+) -> hl.Table:
+    """
+    Count the observed variants and possible variants by exome coverage.
+
+    :param exome_ht: Preprocessed exome Table.
+    :param context_ht: Preprocessed context Table.
+    :param mutation_ht: Preprocessed mutation rate Table.
+    :param possible_file_path: Path to save table with possible variants.
+    :param recompute_possible: Whether to use context Table to recompute the number of possible variants instead of using a precomputed intermediate Table if it exists. Defaults to False.
+    :param impose_high_af_cutoff_upfront: Whether to remove high frequency alleles, defaults to True.
+    :param dataset: Dataset to use when computing frequency index. Defaults to 'gnomad'.
+    :param af_cutoff: Variants with AF above than AF cutoff will be removed, defaults to 0.001.
+    :param kept_context_annotations: Annotations to keep in the context Table.
+    :param kept_exome_annotations: Annotations to keep in the exome Table.
+    :param pops: List of populations. Defaults to `POPS`.
+    :param grouping: Annotations other than 'context', 'ref', 'alt' to group by when counting variants, defaults to 'exome_coverage'.
+    :return: Table with observed variant and possible variant count.
+    """
+    context_ht = fast_filter_vep(context_ht).select(kept_context_annotations)
+    context_ht = context_ht.filter(hl.is_defined(context_ht.exome_coverage))
+    exome_ht = fast_filter_vep(exome_ht).select(kept_exome_annotations)
+
+    context_ht, exome_ht = remove_unnecessary_variants(
+        context_ht, exome_ht, dataset, af_cutoff, impose_high_af_cutoff_upfront
+    )
+
+    if recompute_possible:
+        possible_ht = count_variants(
+            context_ht, additional_grouping=grouping, force_grouping=True
+        )
+        possible_ht = annotate_with_mu(possible_ht, mutation_ht)
+        possible_ht.transmute(possible_variants=possible_ht.variant_count).write(
```

Comment:
Is it necessary to write the possible files here? I remember we usually put checkpoints in pipeline instead of functions.